flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/StateTable.java

```diff
@@ -156,10 +203,23 @@ public boolean isEmpty() {
 	 * @param transformation the transformation function.
 	 * @throws Exception if some exception happens in the transformation function.
 	 */
-	public abstract <T> void transform(
+	public <T> void transform(
 			N namespace,
 			T value,
-			StateTransformationFunction<S, T> transformation) throws Exception;
+			StateTransformationFunction<S, T> transformation) throws Exception {
+		K key = keyContext.getCurrentKey();
+		checkKeyNamespacePreconditions(key, namespace);
+
+		int keyGroup = keyContext.getCurrentKeyGroupIndex();
+		StateMap<K, N, S> stateMap = getMapForKeyGroup(keyGroup);
+
+		if (stateMap == null) {
+			stateMap = createStateMap();
```

Comment:
I think you make a trade-off here between the attempt to save some space, and some code complexity and checking overhead for `null` in the array. Given that the keys are typically uniformly distributed across key-groups, I would assume that no map will stay `null` for long. What I would suggest is to initialize all array positions with maps at the time of array allocation and keep the invariant that all places in the array are strictly non-null. This allows you to drop all null checks on around this array.

What I can forsee is that maybe the reason to have the checks is for followup work that implements the spilling and for spilled key-groups you can have some maps removed from the array? If that is the case, then we could either keep the check or (maybe better) have a special implementation for an empty/spilled map?

Reply:
If you feel like you want to keep it, you might consider writing a helper method for the task of creating and adding maps, because this is repeated code in multiple places.

Reply:
The style of checking `null` is just following the original NestedMapsStateTable. For the followup spill implementation, we just replace the on-heap map with on-disk map, instead of removing the map from the array, so there is no need to set `null` in the array. I agree with you that no map will stay `null` for long if the number of keys is more than the number of key-groups and keys are uniformly distributed, which I think is the most case. So I will initialize all array positions with maps and remove the check.